### PR TITLE
MetaMask E2E test guidelines

### PR DESCRIPTION
**Pull request by cortisiko**
The purpose of this PR is to create a unified document for E2E test guidelines on both platforms.

### CODE REVIEW

**Review by MarioAslau — APPROVED** [review]
*chef's kiss*

File: docs/e2e-testing.md [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+```
+
+⚠️ Test names remains approachable, recommend to avoid using the prefix meaningless 'should'
+
+```javascript
+- should add Bob to the address book
+- should send 1 TST to Bob
+```
+
+❌ Test name should be completely avoided: meaningless `should` prefix, the `and` decreases the readability of the test,making it harder to understand what the test is doing as well as diagnose and fix issues.
```
Comment (sleepytanya) [inline docs/e2e-testing.md:40]: I may be wrong again but is 'test name should be completely avoided' correct? or '"Test names should avoid the meaningless 'should' prefix and the use of 'and', as these decrease readability, making it harder to understand ..." Also space is missing after the comma '...the test,making it...'.

Comment (sleepytanya) [inline docs/e2e-testing.md:40]: 'Test names' instead of singular 'test name' because there are 2 tests.
'...of the test,making it harder..' - missing space after a comma.
'The use of a 'should' prefix and the word 'and' can decrease..' may be?

File: docs/e2e-testing.md [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+### Guidelines
+
+✅ Recommended readable test names
+
+```javascript
+- adds Bob to the address book
+- send 1 TST to Bob
+```
+
+⚠️ Test names remains approachable, recommend to avoid using the prefix meaningless 'should'
```
Comment (sleepytanya) [inline docs/e2e-testing.md:33]: 'remain' instead of 'remains'? 
Is 'meaningless prefix' better than 'prefix meaningless'?
(I thought 'should' is a modal verb?)

File: docs/e2e-testing.md [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+- The future goal is for specific feature teams to manage the necessary test cases for their own features, instead of the core teams tracking each one.
+- Using this organizational strategy eliminates the need to decide at a low level where to place a test, as it is straightforward based on the feature or scenario.
+- This approach aligns with the testing strategy for the mobile app. The mobile team prefers to have feature team tests under the same folder. This reduces friction when switching context.
+
+Example for organization of test files by features and scenarios:
+
+```javascript
+// current test path:
+test/e2e/tests/nft/import-erc1155.spec.js
+// recommanded test path: (consolidate all import tests for different tokens into a single repository)
```
Comment (sleepytanya) [inline docs/e2e-testing.md:82]: Typo in 'recommAnded'.

File: docs/e2e-testing.md [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+test/e2e/tests/tokens/import/import-erc1155.spec.js
+
+// current test path:
+test/e2e/tests/clear-activity.spec.js
+// recommanded test path:
```
Comment (sleepytanya) [inline docs/e2e-testing.md:87]: Typo in 'recommAnded'.

File: docs/e2e-testing.md [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+test/e2e/tests/settings/clear-activity.spec.js
+
+// current test path:
+test/e2e/tests/ppom-blockaid-alert-erc20-approval.spec.js
+// recommanded test path:
```
Comment (sleepytanya) [inline docs/e2e-testing.md:92]: Same typo - 'recommAnded'.

File: docs/e2e/extension-e2e-guidelines.md [hunk 6]
```diff
@@ -1,15 +1,5 @@
...
-
-This document serves as a guideline to approach writing E2E tests so that they provide fast feedback, they become less fragile over time, they are painless to debug, and they are scalable and easy to maintain. The remainder of this document will attempt to explain the suggested approach with relevant examples to help better understand the future direction for tests.
-
-These guidelines aren't meant to be a strict set of rules, they should be thought of as an approach to writing robust end-to-end tests. Use them where they make sense and adapt them as necessary to complement your tests.
-
 ### TL;DR
 
 - Write test names that explain the behaviour of the test
```
Comment (sleepytanya) [inline docs/e2e/extension-e2e-guidelines.md:5]: Would 'test name' be better as it explains the behaviour of singular 'test'?

File: docs/e2e/mobile-e2e-guidelines.md [hunk 7]
```diff
@@ -0,0 +1,550 @@
...
+
+❌ Bad: Hardcoding text strings without retrieving them from the language file, like this:
+
+```javascript
+const elementText = 'Show Hex Data'; // Hardcoded text
+```
+
+## 📌 Naming Convention
+
+Descriptive and clear names are the cornerstone of maintainable code. Vague, cryptic names that leave readers puzzled.
```
Comment (sleepytanya) [inline docs/e2e/mobile-e2e-guidelines.md:53]: May be delete 'that' makes sense?

File: docs/e2e/mobile-e2e-guidelines.md [hunk 8]
```diff
@@ -0,0 +1,550 @@
...
+
+Page objects serve as the building blocks of our test suites, providing a clear and organized representation of the elements and interactions within our application.
+
+When creating page objects, follow these principles to ensure clarity, maintainability, and reusability.
+
+### 🔁 Consistency in Structure
+
+Define a clear structure for your page objects, organizing elements and actions in a logical manner. This makes it easier for other engineers to understand and navigate the code.
+
+✅ Good: Define a clear structure for your page objects, organizing elements and actions in a logical manner. This makes it easy to reuse elements and actions across multiple tests. For example:
```
Comment (sleepytanya) [inline docs/e2e/mobile-e2e-guidelines.md:251]: Does 'Define a clear structure for your page objects, organizing elements and actions in a logical manner.' intentionally duplicates first sentence in line 249?

File: docs/e2e/mobile-e2e-guidelines.md [hunk 9]
```diff
@@ -0,0 +1,550 @@
...
+
+❌ Bad: Implementing interactions directly in test code without using utility functions. For example:
+
+```javascript
+await element(by.id('ELEMENT-STRING')).tap();
+```
+
+### 🗃️ Using Getters for Element Storage
+
+In our page objects, we utilize getters for element storage instead of defining selectors within the constructor. This deliberate choice stems from the behavior of getters, which are evaluated when you access the property, not when you create the object.
```
Comment (sleepytanya) [inline docs/e2e/mobile-e2e-guidelines.md:360]: Inconsistent spelling - behavior / behaviour, would be nice to choose one spelling for all documents.

File: docs/e2e/mobile-e2e-guidelines.md [hunk 10]
```diff
@@ -0,0 +1,550 @@
...
+With the page object in place, writing test specifications becomes more straightforward. The test scripts interact with the application through the page object's interface, improving readability and maintainability.
+
+#### Example Test Case: Adding a Network
+
+```javascript
+import NetworkView from '../../pages/Settings/NetworksView';
+import Assertions from '../../utils/Assertions';
+import Networks from '../../resources/networks.json';
+
+it('should add Gnosis network', async () => {
```
Comment (sleepytanya) [inline docs/e2e/mobile-e2e-guidelines.md:537]: We should avoid using 'should' in test names?

**Review by chrisleewilcox — APPROVED** [review]
I think this is a good start.  This will be a living document and we can update as necessary.

LGTM

File: docs/e2e-testing.md (hunk 1 above)
Comment (cortisiko) [reply docs/e2e-testing.md:40]: Great point, fixed!

File: docs/e2e-testing.md (hunk 3 above)
Comment (cortisiko) [reply docs/e2e-testing.md:82]: fixed!

File: docs/e2e-testing.md (hunk 4 above)
Comment (cortisiko) [reply docs/e2e-testing.md:87]: fixed

File: docs/e2e-testing.md (hunk 5 above)
Comment (cortisiko) [reply docs/e2e-testing.md:92]: fixed

File: docs/e2e/mobile-e2e-guidelines.md (hunk 8 above)
Comment (cortisiko) [reply docs/e2e/mobile-e2e-guidelines.md:251]: oh wonderful eye, this was not intentional. I have updated it.

File: docs/e2e/mobile-e2e-guidelines.md (hunk 9 above)
Comment (cortisiko) [reply docs/e2e/mobile-e2e-guidelines.md:360]: Interesting point Tanya. It is worth noting that there is a slight difference in spelling between Americans and Europeans, even though the words have the same meaning. Depending on who writes the documentation, their spelling might be different. However, I am hesitant to state which spelling is better for words that have the same meaning but are spelled differently.

**Review by sleepytanya — DISMISSED** [review]

File: docs/e2e-testing.md [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+- This approach assigns ownership of E2E testing at the feature team level. Each feature team, being aware of the features they own, finds it easier to understand the existing tests and identify what's missing.
+- The future goal is for specific feature teams to manage the necessary test cases for their own features, instead of the core teams tracking each one.
+- Using this organizational strategy eliminates the need to decide at a low level where to place a test, as it is straightforward based on the feature or scenario.
+- This approach aligns with the testing strategy for the mobile app. The mobile team prefers to have feature team tests under the same folder. This reduces friction when switching context.
+
+Example for organization of test files by features and scenarios:
+
+```javascript
+// current test path:
+test/e2e/tests/nft/import-erc1155.spec.js
```
Comment (chloeYue) [inline docs/e2e-testing.md:81]:
```suggestion
metamask-extension/test/e2e/tests/nft/import-erc1155.spec.js
```

File: docs/e2e-testing.md [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+// recommended test path: (consolidate all import tests for different tokens into a single repository)
+test/e2e/tests/tokens/import/import-erc1155.spec.js
```
Comment (chloeYue) [inline docs/e2e-testing.md:83]:
```suggestion
metamask-extension/test/e2e/tests/tokens/import/import-erc1155.spec.js
```

File: docs/e2e-testing.md [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+
+// current test path:
+test/e2e/tests/clear-activity.spec.js
```
Comment (chloeYue) [inline docs/e2e-testing.md:86]:
```suggestion
metamask-extension/test/e2e/tests/clear-activity.spec.js
```

File: docs/e2e-testing.md [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+// recommended test path:
+test/e2e/tests/settings/clear-activity.spec.js
```
Comment (chloeYue) [inline docs/e2e-testing.md:88]:
```suggestion
metamask-extension/test/e2e/tests/settings/clear-activity.spec.js
```

File: docs/e2e-testing.md [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+
+// current test path:
+test/e2e/tests/ppom-blockaid-alert-erc20-approval.spec.js
```
Comment (chloeYue) [inline docs/e2e-testing.md:91]:
```suggestion
metamask-extension/test/e2e/tests/ppom-blockaid-alert-erc20-approval.spec.js
```

File: docs/e2e-testing.md [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+// recommended test path:
+test/e2e/tests/ppom/ppom-blockaid-alert-erc20-approval.spec.js
```
Comment (chloeYue) [inline docs/e2e-testing.md:93]:
```suggestion
metamask-extension/test/e2e/tests/ppom/ppom-blockaid-alert-erc20-approval.spec.js
```

File: docs/e2e-testing.md [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+Send TST
+Assertion
+```
+
+Here are some examples to remove redundant steps following above guidelines:
+
+1.
+
+```javascript
+// test file: test/e2e/tests/multiple-transactions.spec.js
```
Comment (chloeYue) [inline docs/e2e-testing.md:181]:
```suggestion
// test file: metamask-extension/test/e2e/tests/multiple-transactions.spec.js
```

File: docs/e2e-testing.md [hunk 18]
```diff
@@ -0,0 +1,208 @@
...
+scenario: creates multiple queued transactions and then confirms.
+solution: Ensure tests are different from the tests that line in test/e2e/tests/navigate-transactions.spec.js
```
Comment (chloeYue) [inline docs/e2e-testing.md:183]:
```suggestion
solution: Ensure tests are different from the tests that line in metamask-extension/test/e2e/tests/navigate-transactions.spec.js
```

File: docs/e2e-testing.md [hunk 19]
```diff
@@ -0,0 +1,208 @@
...
+```
+
+2.
+
+```javascript
+// test file: test/e2e/tests/multiple-transactions.spec.js
```
Comment (chloeYue) [inline docs/e2e-testing.md:189]:
```suggestion
// test file: metamask-extension/test/e2e/tests/multiple-transactions.spec.js
```

File: docs/e2e-testing.md [hunk 20]
```diff
@@ -0,0 +1,208 @@
...
+scenario: creates multiple queued transactions, then rejects.
+solution: create multiple transactions using the fixture builder rather than connecting to the test dapp and creating transactions through the UI
+```
+
+3.
+
+```javascript
+// test file: test/e2e/tests/add-account.spec.js
```
Comment (chloeYue) [inline docs/e2e-testing.md:197]:
```suggestion
// test file: metamask-extension/test/e2e/tests/add-account.spec.js
```

File: docs/e2e-testing.md [hunk 21]
```diff
@@ -0,0 +1,208 @@
...
+scenario: should not affect public address when using secret recovery phrase to recover account with non-zero balance
+solution: replace UI steps that build up the application state with the FixtureBuilder
+```
+
+4.
+
+```javascript
+// test file: test/e2e/tests/import-flow.spec.js
```
Comment (chloeYue) [inline docs/e2e-testing.md:205]:
```suggestion
// test file: metamask-extension/test/e2e/tests/import-flow.spec.js
```

File: docs/e2e/extension-e2e-guidelines.md [hunk 22]
```diff
@@ -1,15 +1,5 @@
...
-
-This document serves as a guideline to approach writing E2E tests so that they provide fast feedback, they become less fragile over time, they are painless to debug, and they are scalable and easy to maintain. The remainder of this document will attempt to explain the suggested approach with relevant examples to help better understand the future direction for tests.
-
-These guidelines aren't meant to be a strict set of rules, they should be thought of as an approach to writing robust end-to-end tests. Use them where they make sense and adapt them as necessary to complement your tests.
-
 ### TL;DR
```
Comment (chloeYue) [inline docs/e2e/extension-e2e-guidelines.md:3]: Due to the reorganisation of doc, `TL;DR ` section in this doc should be updated, new items list:

```
### TL;DR

- Manage test size and execution time
- Craft reliable element locators using data-testid
- Avoid using arbitrary delays, leverage wait commands
- Write repeatable tests that produce the same result every time
- Explicitly state assertions
- Enhance test stability with request mocking

```

File: docs/e2e-testing.md [hunk 23]
```diff
@@ -0,0 +1,208 @@
...
+
+Organise tests into folders based on scenarios and features. This means that each type of scenario has its own folder, and each feature team owns one or more folders.
+
+- This approach assigns ownership of E2E testing at the feature team level. Each feature team, being aware of the features they own, finds it easier to understand the existing tests and identify what's missing.
+- The future goal is for specific feature teams to manage the necessary test cases for their own features, instead of the core teams tracking each one.
+- Using this organizational strategy eliminates the need to decide at a low level where to place a test, as it is straightforward based on the feature or scenario.
+- This approach aligns with the testing strategy for the mobile app. The mobile team prefers to have feature team tests under the same folder. This reduces friction when switching context.
```
Comment (chloeYue) [inline docs/e2e-testing.md:75]: I think we should delete this line, as it's in the common doc for extension and mobile

Comment (cortisiko) [reply docs/e2e-testing.md:75]: From Chloe: reword this 

> This approach aligns with the testing strategy for the mobile app. The mobile team prefers to have feature team tests under the same folder. This reduces friction when switching context.

to 

The extension and mobile team both adhere to the same method for organizing test files. This consistency reduces friction when switching context.

Comment (cortisiko) [reply docs/e2e-testing.md:75]: updated!

File: docs/e2e-testing.md (hunk 21 above)
Comment (cortisiko) [reply docs/e2e-testing.md:205]: As discussed in our 1<>1, we opted to omit the root folder in the relative path and specify the path of the e2e tests. Updated this.

File: docs/e2e-testing.md (hunk 20 above)
Comment (cortisiko) [reply docs/e2e-testing.md:197]: As discussed in our 1<>1, we opted to omit the root folder in the relative path and specify the path of the e2e tests. I have updated this.

File: docs/e2e-testing.md (hunk 19 above)
Comment (cortisiko) [reply docs/e2e-testing.md:189]: As discussed in our 1<>1, we opted to omit the root folder in the relative path and specify the path of the e2e tests. I have updated this.

File: docs/e2e-testing.md (hunk 13 above)
Comment (cortisiko) [reply docs/e2e-testing.md:86]: As discussed in our 1<>1, we opted to omit the root folder in the relative path and specify the path of the e2e tests. I have updated this.

File: docs/e2e-testing.md (hunk 12 above)
Comment (cortisiko) [reply docs/e2e-testing.md:83]: As discussed in our 1<>1, we opted to omit the root folder in the relative path and specify the path of the e2e tests. I have updated this.

File: docs/e2e-testing.md (hunk 11 above)
Comment (cortisiko) [reply docs/e2e-testing.md:81]: As discussed in our 1<>1, we opted to omit the root folder in the relative path and specify the path of the e2e tests. I have updated this.

File: docs/e2e-testing.md [hunk 24]
```diff
@@ -0,0 +1,208 @@
...
+- the background environment is locked
+```
+
+❌ Test name should be completely avoided
+
+```javascript
+// in file: test/e2e/tests/add-account.spec.js
+- should be possible to remove an account imported with a private key, but should not be possible to remove an account generated from the SRP imported in onboarding
+
+// in file: test/e2e/tests/lockdown.spec.js
```
[the remark this answers is not in the document]
Comment (cortisiko) [reply docs/e2e-testing.md:62]: As discussed in our 1<>1, we opted to omit the root folder in the relative path and specify the path of the e2e tests. I have updated this.

**Review by chloeYue — APPROVED** [review]
LGTM
